foqus_lib/framework/surrogate/keras_nn.py

```diff
@@ -0,0 +1,465 @@
+
+        # xlabels = xdata.columns.tolist()  # set labels as a list (default) from pandas
+        # zlabels = zdata.columns.tolist()  #    is a set of IndexedDataSeries objects
+
+        xlabels = str(self.options["input_labels"].value).split(",")
+        zlabels = str(self.options["output_labels"].value).split(",")
+
+        # tried the below code that didn't work
+        xdata = data.loc[:, xlabels]  # there are 6 input variables/columns
+        zdata = data.loc[:, zlabels]  # the rest are output variables/columns
+
+        xdata_bounds = {i: (xdata[i].min(), xdata[i].max()) for i in xdata}  # x bounds
+        zdata_bounds = {j: (zdata[j].min(), zdata[j].max()) for j in zdata}  # z bounds
+
+        # normalize data using Linear form
+        # users can normalize with any allowed form # manually, and then pass the
+        # appropriate flag to FOQUS from the allowed list:
+        # ["Linear", "Log", "Power", "Log 2", "Power 2"] - see the documentation for
+        # details on the scaling formulations
+        xmax, xmin = xdata.max(axis=0), xdata.min(axis=0)
+        zmax, zmin = zdata.max(axis=0), zdata.min(axis=0)
+        xdata, zdata = np.array(xdata), np.array(zdata)
+        for i in range(len(xdata)):
+            for j in range(len(xlabels)):
+                xdata[i, j] = (xdata[i, j] - xmin[j]) / (xmax[j] - xmin[j])
+            for j in range(len(zlabels)):
+                zdata[i, j] = (zdata[i, j] - zmin[j]) / (zmax[j] - zmin[j])
+
+        model_data = np.concatenate(
+            (xdata, zdata), axis=1
+        )  # Keras requires a Numpy array as input
+
+        # define x and z data, not used but will add to variable dictionary
+        xdata = model_data[:, :-2]
+        zdata = model_data[:, -2:]
+
+        # method to create model
+        def create_model(data):
+
+            inputs = tf.keras.Input(shape=(np.shape(data)[1],))  # create input layer
+
+            layers = keras_nn(  # define the rest of network using our custom class
+                # add our 4 options
+                input_labels=xlabels,
+                output_labels=zlabels,
+                input_bounds=xdata_bounds,
+                output_bounds=zdata_bounds,
+                normalized=True,
+                normalization_form="Linear",
+            )
+
+            outputs = layers(inputs)  # use network as function outputs = f(inputs)
+
+            model = tf.keras.Model(inputs=inputs, outputs=outputs)  # create model
+
+            model.compile(loss="mse", optimizer="RMSprop", metrics=["mae", "mse"])
+
+            model.fit(
+                xdata, zdata, epochs=epochs_count, verbose=verbose_setting
+            )  # train model
+
+            return model
+
+        # create model
+        model = create_model(xdata)
+        model.summary()
+        # self.msgQueue.put(model)
+
+        self.msgQueue.put("Training complete")
+        # save model as H5
+        model.save("keras_nn.h5")
```

Comment:
I'd recommend saving the models to the `user_ml_ai_models` folder in the working directory. When testing locally, `model.save("user_ml_ai_models/keras_nn.h5")` seemed to work.